viewer/mlab-view

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/python
+# Copyright 2014 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Summary:
+  mlab-view provides a limited web server to view the collectd-web CGI scripts.
+
+  While mlab-view is running, an operator can only access the CGI scripts using
+  an SSH tunnel. Direct web connections are refused with "403 Forbidden."
+
+  mlab-view should only be used for debugging and node triage. When a node is
+  in normal operation, mlab-view should not be running. mlab-view guarantees
+  that the web server is shutdown when the operator's ssh connection has
+  disconnected.
+
+Example usage (from operator's workstation):
+
+  $ ssh -L 8088:utility.mlab.<hostname>:8088 \\
+      -p806 mlab_utility@<hostname> sudo mlab-view
+
+  Operator directs their browser to:
+
+      http://localhost:8088/
+
+What's happening?
+
+ * SSH binds to port 8088 on localhost.
+ * SSH login to <hostname> as the mlab_utility user and runs the command:
+
+     "sudo mlab-view".
+
+ * mlab-view starts a web server that listens on port 8088.
+ * The web server is configured to reject connections from any address other
+   than the slice IP.
+ * When ssh detects a connection to localhost:8088, the remote end attempts a
+   connection to:
+
+      "utility.mlab.<hostname>:8088"
+
+   To the server, this connection appears to come from the slice IP.
+ * SSH continues to forward all traffic back and forth from localhost:8088
+   through the ssh connection to the remote web server.
+"""
+
+import os
+import signal
+import subprocess
+import sys
+import syslog
+import time
+
+# Third-party modules.
+import netifaces
+
+
+ACCESS_CONF = '/usr/share/collectd-mlab/access.conf'
+HTTPD_PID = '/etc/httpd/run/httpd.pid'
```

Comment:
That's right. Apache writes that file but does not clean it up. The standard init scripts help with this cleanup, but since we are not using the init system here the cleanup is handled by mlab-view.